pkg/controller/common/tracing/apmclientgo/client.go

```diff
@@ -0,0 +1,178 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package apmclientgo
+
+import (
+	"io"
+	"net/http"
+	"strings"
+	"sync/atomic"
+	"unsafe"
+
+	"github.com/pkg/errors"
+	"go.elastic.co/apm/module/apmhttp/v2"
+	"go.elastic.co/apm/v2"
+)
+
+// WrapRoundTripper returns a http.Roundtripper wrapping r, reporting each
+// request as a span to Elastic APM, if the request's context contains a sampled transaction
+// Allows an optional default transaction to be configured for requests where context cannot be controlled
+// e.g. client-go's cache management
+func WrapRoundTripper(r http.RoundTripper, o ...ClientOption) http.RoundTripper {
+	if r == nil {
+		r = http.DefaultTransport
+	}
+	rt := &roundTripper{r: r}
+	// apply any client options
+	for _, o := range o {
+		o(rt)
+	}
+	return rt
+}
+
+type roundTripper struct {
+	r           http.RoundTripper
+	defaultTxFn func() *apm.Transaction
+}
+
+func (r roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	ctx := req.Context()
+	tx := apm.TransactionFromContext(ctx)
+	if tx == nil && r.defaultTxFn != nil {
+		tx = r.defaultTxFn()
+		defer tx.End()
```

Comment:
It may panic here if there is no `tracer`:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x472950e]

goroutine 30 [running]:
go.elastic.co/apm/v2.(*Transaction).End(0x0)
        /Users/michael/go/pkg/mod/go.elastic.co/apm/v2@v2.0.0/transaction.go:277 +0x2e
github.com/elastic/cloud-on-k8s/pkg/controller/common/tracing/apmclientgo.roundTripper.RoundTrip({{0x5b1a720?, 0xc0002437c0?}, 0xc000449fd0?}, 0xc00060d400)
        /Users/michael/go/src/github.com/elastic/cloud-on-k8s/pkg/controller/common/tracing/apmclientgo/client.go:48 +0x8b2
golang.org/x/oauth2.(*Transport).RoundTrip(0xc000098840, 0xc00060d300)
[...]
```

Maybe we should set the wrapper only if `tracer` is non `nil`, for example:

```golang
	// set up APM  tracing if configured
	var tracer *apm.Tracer
	if viper.GetBool(operator.EnableTracingFlag) {
		tracer = tracing.NewTracer("elastic-operator")
		// set up APM tracing for client-go
		cfg.Wrap(tracing.ClientGoTransportWrapper(
			apmclientgo.WithDefaultTransaction(tracing.ClientGoCacheTx(tracer)),
		))
	}
```